soccer/radio/NetworkRadio.cpp

```diff
@@ -0,0 +1,187 @@
+#include "NetworkRadio.hpp"
+#include <Utils.hpp>
+#include "Geometry2d/Util.hpp"
+#include "status.h"
+
+using namespace boost::asio;
+using ip::udp;
+
+NetworkRadio::NetworkRadio(int server_port, int robot_port)
+    : _socket(_context, udp::endpoint(udp::v4(), server_port)),
+      _server_port{server_port},
+      _robot_port{robot_port} {}
+
+NetworkRadio::~NetworkRadio() {}
+
+bool NetworkRadio::open() {
+    startReceive();
+    return true;
+}
+
+void NetworkRadio::startReceive() {
+    // Set a receive callback
+    _socket.async_receive_from(
+        boost::asio::buffer(_recv_buffer), _robot_endpoint,
+        [this] (const boost::system::error_code& error, std::size_t num_bytes) {
+            receivePacket(error, num_bytes);
+        });
+}
+
+bool NetworkRadio::isOpen() const {
+    return _socket.is_open();
+}
+
+void NetworkRadio::send(Packet::RadioTx& packet) {
+    // Get a list of all the IP addresses this packet needs to be sent to
+    for (int robot_idx = 0; robot_idx < packet.robots_size(); robot_idx++) {
+        const Packet::Control& control = packet.robots(robot_idx).control();
+        uint32_t robot_id = packet.robots(robot_idx).uid();
+
+        // Build the control packet for this robot.
+        std::array<uint8_t, rtp::HeaderSize + sizeof(rtp::RobotTxMessage)> forward_packet;
+
+        rtp::Header* header = reinterpret_cast<rtp::Header*>(&forward_packet[0]);
+        header->port = rtp::PortType::CONTROL;
+        header->address = rtp::BROADCAST_ADDRESS;
+        header->type = rtp::MessageType::CONTROL;
+
+        size_t offset = sizeof(rtp::Header);
+        rtp::RobotTxMessage* msg = reinterpret_cast<rtp::RobotTxMessage*>(
+                &forward_packet[offset]);
+
+        // Set fields of the control message.
+        msg->uid = robot_id;
+
+        msg->message.controlMessage.bodyX = static_cast<int16_t>(
+            control.xvelocity() * rtp::ControlMessage::VELOCITY_SCALE_FACTOR);
+        msg->message.controlMessage.bodyY = static_cast<int16_t>(
+            control.yvelocity() * rtp::ControlMessage::VELOCITY_SCALE_FACTOR);
+        msg->message.controlMessage.bodyW = static_cast<int16_t>(
+            control.avelocity() * rtp::ControlMessage::VELOCITY_SCALE_FACTOR);
+        msg->message.controlMessage.dribbler =
+            clamp(static_cast<uint16_t>(control.dvelocity()) * 2, 0, 255);
```

Comment:
*2?

Reply:
Good question. Short answer is that this is copy pasted from USBRadio, which is another issue, but I didn't want to refactor the entire radio stack because I thought this was more urgent than it actually was.